1-Draft/RFCNNN-WindowsPS-Compatibility-Utilities.md

```diff
@@ -0,0 +1,106 @@
+So a new IPC-specific parameter set will be added to New-PSSession:  `New-PSSession [-Name <String>] -ProcessId <Int32>`
+Here is the pseudo-code that shows the essence of proposed implementation of a new parameter set in New-PSSession:
+```csharp
+string WindowsPS_AppDomainName = GetPSHostProcessInfoCommand.GetAppDomainNamesFromProcessId(WindowsPS_ProcessId);
+NamedPipeConnectionInfo connectionInfo = new NamedPipeConnectionInfo(WindowsPS_ProcessId, WindowsPS_AppDomainName);
+TypeTable typeTable = TypeTable.LoadDefaultTypeFiles();
+RemoteRunspace remoteRunspace = RunspaceFactory.CreateRunspace(connectionInfo, this.Host, typeTable) as RemoteRunspace;
+remoteRunspace.Name = NamedPipeRunspaceName;
+remoteRunspace.ShouldCloseOnPop = true
+try
+{
+    remoteRunspace.Open();
+}
+catch (RuntimeException e)
+{
+    // handle connection errors
+}
+PSSession remotePSSession = new PSSession(remoteRunspace);
+this.RunspaceRepository.Add(remotePSSession);
+WriteObject(remotePSSession);
+```
+
+### Lifetime of remote Windows PowerShell process and module
+
+Overall RFC goal is to have farmiliar 'local module' user experience even though actuall operations are working with a module in a remote process. This drives following:
+
+1. One remote Windows PS process corresponds to one local PS Core process. E.g. if a same user creates 2 local PS Core processes and loads 'WindowsPS-only' module in each one, this will result in creating 2 Windows PS processes and loading the actuall module in each one of them.
+2. Lifetime of the module on remote side (Windows PS) should be the same as on local side (PS Core). This is important because some modules save their global state in-between command calls.
+   * Local Load-Module for 'WindowsPS-only' module will have same behaviour/effect in remote Windows PS process as it would have been done locally on an compartible module. In addition, this will create `WindowsPS Compatibility` Windows PS process if one does not already exist.
+   * Local Remove-Module for 'WindowsPS-only' module will have same behaviour/effect in remote Windows PS process as it would have been done locally on an compartible module. In addition, this will remove `WindowsPS Compatibility` Windows PS process if it is no longer needed.
+3. When removing a module, there is an `OnRemove` event on the module that will execute. This event allows Compat proxy module to react to being removed and perform cleanup of remote process if necessary.
+4. PS process exit does Not perform gracefull cleanup of modules, so it needs to be handled separately. There is an event that allows to react to the closing of the PowerShell process (to do remote Compat process cleanup): `Register-EngineEvent -SourceIdentifier ([System.Management.Automation.PsEngineEvent]::Exiting) -Action $OnRemoveScript`

```

Comment:
My main concern is a dangling WindowsPS process after PowerShell Core exits.  I don't think there is an OS way to guarantee a child process exits after the parent process exits.  That was one nice thing about WinRM service, as it could notify notify parent if child exited and terminate the child process if the parent process abruptly exited.

But I don't think we need a remoting event, and instead can use the Process.Exited event.  The child could monitor the parent and if the parent (PowerShell Core) exits unexpectedly then the child process can exit.  Or we could have some other clean up mechanism to ensure child processes are terminated at some point.

Reply:
Good point. Monitoring ParentProcess.Exited from child should work.